File: rust/axum/src/modules/auth/controller.rs
use axum::{Json, http::StatusCode};
use serde::{Deserialize, Serialize};
use crate::common::errors::base::BaseError;
use crate::common::errors::bad_request::bad_request_error;
use crate::common::messages::error::ErrorMessages;
use validator::Validate;

use super::service::AuthService;
use super::dto::{LoginAuthDto, LoginAuthClientOutputDto, LoginAuthManagerOutputDto};

#[derive(Debug)]
pub struct AuthController {
}

impl AuthController {
    pub async fn login_client(
        Json(dto): Json<LoginAuthDto>,
    ) -> Result<Json<LoginAuthClientOutputDto>, BaseError> {
        if let Err(errors) = dto.validate() {
            return Err(bad_request_error(ErrorMessages::BAD_REQUEST, Some(errors)));
        }

        match AuthService::login_client(dto).await {
            Ok(result) => Ok(Json(result)),
            Err(err) => Err(err),
        }
    }

    pub async fn login_manager(
        Json(dto): Json<LoginAuthDto>,
    ) -> Result<Json<LoginAuthManagerOutputDto>, BaseError> {
        if let Err(errors) = dto.validate() {
            return Err(bad_request_error(ErrorMessages::BAD_REQUEST, Some(errors)));
        }

        match AuthService::login_manager(dto).await {
            Ok(result) => Ok(Json(result)),
            Err(err) => Err(err),
        }
    }
}

File: go/fiber_orm/internal/modules/auth/controller.go
package auth

import (
	"tcc/internal/common/errors"
	"tcc/internal/common/helper"
	"tcc/internal/common/messages"
	"tcc/internal/modules/auth/dto"

	"github.com/gofiber/fiber/v2"
)

type AuthController struct{
	service AuthService
}

func (c *AuthController) LoginClient(ctx *fiber.Ctx) error {
	var dto dto.LoginAuthDto
	if err := ctx.BodyParser(&dto); err != nil {
		badRequest := errors.BadRequestError(messages.ErrorMessages.BadRequest, err.Error())
		return ctx.Status(badRequest.StatusCode).JSON(badRequest)
	}
	
	if err := helper.Validate(dto); err != nil {
		badRequest := errors.BadRequestError(messages.ErrorMessages.BadRequest, err...)
		return ctx.Status(badRequest.StatusCode).JSON(badRequest)
	}

	response, err := c.service.LoginClient(dto)
	if err != nil {
		return ctx.Status(err.StatusCode).JSON(err)
	}

	return ctx.Status(fiber.StatusOK).JSON(response)
}

func (c *AuthController) LoginManager(ctx *fiber.Ctx) error {
	var dto dto.LoginAuthDto
	if err := ctx.BodyParser(&dto); err != nil {
		badRequest := errors.BadRequestError(messages.ErrorMessages.BadRequest, err.Error())
		return ctx.Status(badRequest.StatusCode).JSON(badRequest)
	}
	
	if err := helper.Validate(dto); err != nil {
		badRequest := errors.BadRequestError(messages.ErrorMessages.BadRequest, err...)
		return ctx.Status(badRequest.StatusCode).JSON(badRequest)
	}

	response, err := c.service.LoginManager(dto)
	if err != nil {
		return ctx.Status(err.StatusCode).JSON(err)
	}

	return ctx.Status(fiber.StatusOK).JSON(response)
}